### PR TITLE
Add error messages for missing tf and torch imports

### PR DESCRIPTION
**Pull request by avnishn**
## Why are these changes needed?

RLlib gives difficult to understand bug messages when tensorflow/torch aren't installed but experiments are run using these frameworks.

Add a check to see that the necessary nn library is installed before running experiments involving the nn library

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

This pr isn't tested, because I'm not sure how to remove tensorflow or torch from sys.path during my tests and then add it back in later. I suppose I could create a new test target that doesn't have both of the NN libraries and run tests within there, but that seems like unnecessary overhead. @sven1977 @gjoliver could you please advise?

### CODE REVIEW

**Comment by gjoliver** [comment]
cool! I have a high level comment, can you add a module in rllib/util/, something like config_validation.py.
and then create a top level util func,

def validate_trainer_cfg(cfg):

which calls into a bunch of per-rule / per-check functions, such as:

def validate_tf_torch_for_framework(framework: str):
  if framework in ["tf", "tf2", "tfe"]:
    tf, _, _ = try_import_tf()
  else if ...

and move your logics there?
that way, we keep trainer relatively clean, and things we actually check are grouped by functions, so we can easily add a lot more later.
what do you think?

also noticed that I think we should only try_import_tf() if the user actually specified tf as the framework.
otherwise, we will be importing tf and torch at the same time all the time, which may be unnecessary.

File: rllib/agents/trainer.py [hunk 1]
```diff
@@ -661,6 +662,52 @@ def setup(self, config: PartialTrainerConfigDict):
...
+                     "installation found. You can install tensorflow via "
+                     "pip: pip install tensorflow"))
+        elif framework == "torch":
+            if not torch:
+                raise ImportError(
+                    ("torch, was the specified as the 'framework' inside "
+                     "of your config dictionary, however, there was no "
+                     "installation found. You can install torch via "
+                     "pip: pip install torch"))
+        else:
```
Comment (sven1977) [inline rllib/agents/trainer.py:681]: This is actually ok. Some users use neither TF nor Torch with RLlib, so they specify e.g. framework=None.

**Review by sven1977 — COMMENTED** [review]
Great fix, thanks @avnishn !

Two things:
- I'm with Jun on starting to move all the pre-check logic that you will be working on in the near future into the rllib/utils/ folder. Maybe create a new module in there, or use `debug.py` or create a new sub-folder, up to you. There already exist some related modules in `rllib/utils/error.py` (Env-check related error messages).
- framework=None (or any other value) is actually ok and we shouldn't produce an error. This is why the bare-metal policy test case is failing.

File: rllib/agents/trainer.py (hunk 1 above)
Comment (avnishn) [reply rllib/agents/trainer.py]: fixed

File: rllib/agents/trainer.py [hunk 2]
```diff
@@ -27,7 +27,8 @@
 from ray.rllib.models import MODEL_DEFAULTS
```
Comment (richardliaw) [inline rllib/agents/trainer.py:27]: add a test?

File: rllib/agents/trainer.py [hunk 3]
```diff
@@ -661,7 +662,7 @@ def setup(self, config: PartialTrainerConfigDict):
         # be a partial config dict with the class' default).
         self.config = self.merge_trainer_configs(self._default_config, config,
                                                  self._allow_unknown_configs)
-
+        validate_config(self.config)
```
Comment (richardliaw) [inline rllib/agents/trainer.py:665]: validate_framework_config

File: rllib/utils/debug_utils.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
+"""Module for validating and checking various RLlib components."""
+import logging
+from typing import Dict
+
+from ray.rllib.utils.framework import try_import_tf, try_import_torch
+
+logger = logging.getLogger(__name__)
+
+
+def validate_config(config: Dict) -> None:
```
Comment (richardliaw) [inline rllib/utils/debug_utils.py:10]: add a test

File: rllib/BUILD [hunk 5]
```diff
@@ -1354,6 +1354,13 @@ py_test(
     srcs = ["tests/test_dependency_torch.py"]
 )
 
+py_test(
+    name = "tests/test_nn_framework_import_errors",
+    tags = ["team:ml", "tests_dir", "tests_dir_D"],
+    size = "small",
+    srcs = ["teststest_nn_framework_import_errors.py"]
```
Comment (richardliaw) [inline rllib/BUILD:1361]: 1. the directory is wrong right? cc @sven1977 
2. there seems to be a typo in srcs?

File: rllib/agents/trainer.py [hunk 6]
```diff
@@ -1615,6 +1588,80 @@ def make_callbacks():
...
+                        ("torch, was the specified as the 'framework' inside "
+                         "of your config dictionary, however, there was no "
+                         "installation found. You can install torch via "
+                         "pip: pip install torch"))
+
+        def resolve_tf_settings():
+            """Check and resolve DL framework settings.
+
+                Tf-eager (tf2|tfe), possibly with tracing set to True.
+                Recommendsetting tracing to True for speedups.
```
Comment (richardliaw) [inline rllib/agents/trainer.py:1632]: typos here?

File: rllib/agents/trainer.py [hunk 7]
```diff
@@ -1615,6 +1588,80 @@ def make_callbacks():
...
+                if not (_tf1 or _tf):
+                    raise ImportError((
+                        "TensorFlow, was the specified as the 'framework' "
+                        "inside of your config dictionary, however, there was "
+                        "no installation found. You can install tensorflow via"
+                        " pip: pip install tensorflow"))
+            elif framework == "torch":
+                if not _torch:
+                    raise ImportError(
+                        ("torch, was the specified as the 'framework' inside "
```
Comment (richardliaw) [inline rllib/agents/trainer.py:1623]: hmm, why is there a comma here?

Comment (richardliaw) [reply rllib/agents/trainer.py:1623]: btw this grammar doesn't read correctly -- could you fix this?

**Review by richardliaw — CHANGES_REQUESTED** [review]
seems like there are some typos. Fundamentally looks good but please clean up?

File: rllib/BUILD [hunk 8]
```diff
@@ -1354,6 +1354,13 @@ py_test(
     srcs = ["tests/test_dependency_torch.py"]
 )
 
+py_test(
+    name = "tests/test_nn_framework_import_errors",
+    tags = ["team:ml", "tests_dir", "tests_dir_D"],
```
Comment (richardliaw) [inline rllib/BUILD:1359]: this isnt the right test dir right?

**Review by richardliaw — COMMENTED** [review]
Looks good, ping when tests pass?

Please make sure that test_nn_framework is actually being run (link me to the tests)

File: rllib/agents/trainer.py [hunk 9]
```diff
@@ -1615,6 +1588,74 @@ def make_callbacks():
                            cls._allow_unknown_subkeys,
                            cls._override_all_subkeys_if_type_changes)
 
+    @staticmethod
+    def validate_framework(config: PartialTrainerConfigDict) -> None:
+        """Validates the config dictionary.
+
+        Args:
+            config: The config dictionary to be validated.
+
```
Comment (sven1977) [inline rllib/agents/trainer.py:1597]:
```suggestion
```

File: rllib/agents/trainer.py [hunk 10]
```diff
@@ -1615,6 +1588,74 @@ def make_callbacks():
                            cls._allow_unknown_subkeys,
                            cls._override_all_subkeys_if_type_changes)
 
+    @staticmethod
+    def validate_framework(config: PartialTrainerConfigDict) -> None:
+        """Validates the config dictionary.
```
Comment (sven1977) [inline rllib/agents/trainer.py:1593]:
```suggestion
        """Validates the config dictionary wrt the framework settings.
```

File: rllib/agents/trainer.py [hunk 11]
```diff
@@ -1615,6 +1588,74 @@ def make_callbacks():
...
+                if not (_tf1 or _tf):
+                    raise ImportError((
+                        "TensorFlow was specified as the 'framework' "
+                        "inside of your config dictionary. However, there was "
+                        "no installation found. You can install tensorflow via"
+                        " pip: pip install tensorflow"))
+            elif framework == "torch":
+                if not _torch:
+                    raise ImportError(
+                        ("torch was specified as the 'framework' inside "
```
Comment (sven1977) [inline rllib/agents/trainer.py:1623]:
```suggestion
                        ("PyTorch was specified as the 'framework' inside "
```

File: rllib/agents/trainer.py [hunk 12]
```diff
@@ -1615,6 +1588,74 @@ def make_callbacks():
...
+                         "of your config dictionary. However, there was no "
+                         "installation found. You can install torch via "
+                         "pip: pip install torch"))
```
Comment (sven1977) [inline rllib/agents/trainer.py:1626]:
```suggestion
                         "`pip install torch`"))
```

File: rllib/agents/trainer.py [hunk 13]
```diff
@@ -1615,6 +1588,74 @@ def make_callbacks():
...
+        def check_if_correct_nn_framework_installed():
+            """Check if tf/torch experiment is running and tf/torch installed.
+            """
+            if framework in tf_valid_frameworks:
+                if not (_tf1 or _tf):
+                    raise ImportError((
+                        "TensorFlow was specified as the 'framework' "
+                        "inside of your config dictionary. However, there was "
+                        "no installation found. You can install tensorflow via"
+                        " pip: pip install tensorflow"))
```
Comment (sven1977) [inline rllib/agents/trainer.py:1619]:
```suggestion
                        "no installation found. You can install TensorFlow via "
                        "`pip install tensorflow`"))
```

**Review by sven1977 — APPROVED** [review]
Looks great! Thanks for the PR @avnishn .

**Review by richardliaw — APPROVED** [review]

**Pull request merged**